doc/media-settings/Media-based-Port-settings.md

```diff
@@ -0,0 +1,313 @@
+                    "idriver": [
+
+                        "0x2",
+
+                        "0x2"
+
+                    ],
+
+                    "ipredriver": [
+
+                        "0x2",
+
+                        "0x2"
+
+                    ]
+
+                }
+
+            },
+
+            "40GBASE-SR4": {
+
+                "Default": {
+
+                    "preemphasis": [
+
+                        "0x4321",
+
+                        "0x4321"
+
+                    ],
+
+                    "idriver": [
+
+                        "0x2",
+
+                        "0x3"
+
+                    ],
+
+                    "ipredriver": [
+
+                         "0x3",
+
+                         "0x1"
+
+                    ]
+
+                },
+
+                "DELL-00-11-22": {
+
+                    "preemphasis": [
+
+                        "0x1311",
+
+                        "0x321c",
+
+                    ],
+
+                    "idriver": [
+
+                        "0x1",
+
+                        "0x2",
+
+                    ],
+
+                    "ipredriver": [
+
+                        "0x2",
+
+                        "0x1",
+
+                    ]
+
+                }
+
+            },
+
+            "40GBASE-CR4-3M": {
+
+                "Default": {
+
+                    "preemphasis": [
+
+                        "0x1021",
+
+                        "0x3021",
+
+                    ],
+
+                    "idriver": [
+
+                        "0x4",
+
+                        "0x4"
+
+                    ],
+
+                    "ipredriver": [
+
+                        "0x4",
+
+                        "0x4"
+
+                    ]
+
+                },
+
+                "DELL": {
+
+                    "preemphasis": [
+
+                        "0x1311",
+
+                        "0x1312
+
+                    ],
+
+                    "idriver": [
+
+                        "0x2",
+
+                        "0x2"
+
+                    ],
+
+                    "ipredriver": [
+
+                        "0x2",
+
+                        "0x2"
+
+                    ]
+
+                }
+
+            }
+
+        }
+
+    }
+
+}
+
+![](https://github.com/dgsudharsan/SONiC/blob/hld_media/doc/media-settings/event_flow.png)
+
+Flow:
+=====
+
+              When a media is detected in xcvrd daemon, it constructs the media key identifier string as discussed above and searches the media_settings.json file. On finding a matching media key entry the vendor key is constructed and if found the key value pairs are fetched. The xcvrd daemon notifies these key value pairs along with the alias port to portsorch. The portsorch task will convert the alias to SAI object Id and notifies the syncd by doing port attribute set of corresponding settings. Syncd invokes the SAI port set those attribute with specified values. The SAI implementation then programs the attribute into the hardware.
```

Comment:
would you please elaborate on the interaction between the xcvrd and portsorch? like to way to send the notification, the notification message outline?

Reply:
The data will be sent based on notifications. xcvrd will have a NotificationProducer which will send notifications on transceiver detection. The message will contain the following key value pairs 1) Portname = 'logical Port name' on which transceiver detected.
2) Each key value pairs defined in the media_settings file corresponding to the media_type, vendor and port.
The portsorch on receiving this message will convert the logical port name to sai port id and map each key value pair to corresponding attribute?
An example of notification as seen from redis consumer
1) "message"
2) "MEDIANOTIFICATION"
3) "[\"media_change\",\"\",\"port\",\"Ethernet12\",\"ipredriver\",\"0x1\",\"preemphasis\",\"0x5678\",\"idriver\",\"0x1\"]"
1) "message"
2) "MEDIANOTIFICATION"
3) "[\"media_change\",\"\",\"port\",\"Ethernet20\",\"ipredriver\",\"0x1\",\"preemphasis\",\"0x1234\",\"idriver\",\"0x1\"]"
1) "message"
2) "MEDIANOTIFICATION"
3) "[\"media_change\",\"\",\"port\",\"Ethernet116\",\"ipredriver\",\"0x1\",\"preemphasis\",\"0x9abc\",\"idriver\",\"0x1\"]"